src/Faithlife.Reflection/DtoInfo.cs

```diff
@@ -108,6 +108,65 @@ public sealed class DtoInfo<T> : IDtoInfo
 		[return: NotNull]
 		public T CreateNew() => m_lazyCreateNew.Value()!;
 
+		/// <summary>
+		/// Creates a new instance of the DTO.
+		/// </summary>
+		/// <param name="argsAndProps">A collection of DTO constructor arguments and mutable DTO properties/fields with which to create and populate the new DTO instance.</param>
+		/// <exception cref="ArgumentException">No combination of constructor and mutable properties/fields is compatible with the passed items.</exception>
+		[return: NotNull]
+		public T CreateNew(IEnumerable<(string Name, object? Value)> argsAndProps)
+		{
+			try
+			{
+				// Try for a constructor that accepts all passed arguments
+				string[] argNames = argsAndProps.Select(arg => arg.Name).ToArray();
+				object?[] argValues = argsAndProps.Select(arg => arg.Value).ToArray();
+				var flags = BindingFlags.Public | BindingFlags.Instance | BindingFlags.CreateInstance;
+				return (T) typeof(T).InvokeMember(typeof(T).Name, flags, null, null, argValues, null, null, argNames);
+			}
+			catch (MissingMethodException)
```

Comment:
For what it's worth, throwing and catching exceptions as part of normal code flow is bad for performance and generally not a good practice.

Reply:
Ah, the Python community got me into a bad habit in that way, then, haha.

Now that I'm looking for alternatives, though, I'm not finding any other reflection methods that can make intelligent matches using parameter names like [`InvokeMember`](https://docs.microsoft.com/en-us/dotnet/api/system.type.invokemember?view=netcore-3.1) can. [`GetMember`](https://docs.microsoft.com/en-us/dotnet/api/system.type.getmember?view=netcore-3.1), [`GetMethod`](https://docs.microsoft.com/en-us/dotnet/api/system.type.getmethod?view=netcore-3.1), and [`GetConstructor`](https://docs.microsoft.com/en-us/dotnet/api/system.type.getconstructor?view=netcore-3.1) all require knowledge of what order to put the parameters in, and when I [search the .NET docs for "namedParameters"](https://docs.microsoft.com/en-us/search/?terms=namedParameters&scope=.NET), all that comes up is various incarnations of `InvokeMember`.

Should I go ahead and finish that `DtoConstructor` class and add functionality to test a constructor to find if it's compatible given parameter names and types? That way we could find out whether `InvokeMethod` will work before calling it, so we wouldn't have to catch an exception. Or would that be too much reinventing of the wheel?

Reply:
I didn't even know `InvokeMember` could do that!

I've pushed my work to this PR: https://github.com/Faithlife/FaithlifeReflection/pull/7

It maintains a dictionary of parameter indices to get the order right.

It doesn't expose the constructors in the API, but it does maintain a list of compatible constructors ("creators") internally.

Reply:
Ah yeah, that'll do the trick! So no need to worry about constructors with parameter names that don't match the properties they affect?

Come to think of it, I suppose I don't need to assume we don't know _anything_ about the DTOs this class will encounter—given their purpose of "simply" transferring data, it's probably pretty safe to assume there's no reason for their constructors to do anything but directly assign values to properties, in which case there's no reason to expect parameter names to be any different from property names (except in letter casing).

Reply:
Should I go ahead and close this PR? 👍🏼

Reply:
> constructors with parameter names that don't match the properties they affect

That PR rejects any constructor with a parameter name that doesn't match a property. But we do assume that if the names match, it just sets the property.

> Should I go ahead and close this PR?

You bet. Thanks!